### PR TITLE
8140442: Add getOutermostTypeElement to javax.lang.model utility class

### PR DESCRIPTION
**Pull request by jddarcy**
Please review this change to add a getOutermostTypeElement method to javax.lang.model.Elements along with the corresponding CSR https://bugs.openjdk.java.net/browse/JDK-8271903.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8140442](https://bugs.openjdk.java.net/browse/JDK-8140442): Add getOutermostTypeElement to javax.lang.model utility class

### Reviewers
 * [Jan Lahoda](https://openjdk.java.net/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to bc1de3f0968a3f1b00b20f5094b7d2bc494de438

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5007/head:pull/5007` \
`$ git checkout pull/5007`

Update a local copy of the PR: \
`$ git checkout pull/5007` \
`$ git pull https://git.openjdk.java.net/jdk pull/5007/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5007`

View PR using the GUI difftool: \
`$ git pr show -t 5007`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5007.diff">https://git.openjdk.java.net/jdk/pull/5007.diff</a>

</details>

### CODE REVIEW

File: src/java.compiler/share/classes/javax/lang/model/util/Elements.java [hunk 1]
```diff
@@ -533,6 +533,67 @@ default ModuleElement getModuleOf(Element e) {
...
+        // sort of class or interface. Check to see if the element is
+        // a top-level type; if so, return it. Otherwise, keep going
+        // up the enclosing element chain until a top-level type is
+        // found.
+        default -> {
+            Element enclosing = e;
+            // This implementation is susceptible to infinite loops
+            // for misbehaving element implementations.
+            while (true) {
+                // Conceptual instanceof TypeElment check. If the
```
Comment (lahodaj) [inline src/java.compiler/share/classes/javax/lang/model/util/Elements.java:581]: Nit:
```suggestion
                // Conceptual instanceof TypeElement check. If the
```

File: test/langtools/tools/javac/processing/model/util/elements/TestOutermostTypeElement.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    private class OuterScanner extends ElementScanner<Void, Elements> {
+        private TypeElement expectedOuter;
+        public OuterScanner(TypeElement expectedOuter) {
+            this.expectedOuter = expectedOuter;
+        }
+
+        @Override
+        public Void scan(Element e, Elements elts) {
+            // System.out.println(e.getSimpleName() + " " + e.getKind());
```
Comment (lahodaj) [inline test/langtools/tools/javac/processing/model/util/elements/TestOutermostTypeElement.java:78]: Nit:
```suggestion
```

**Review by lahodaj — APPROVED** [review]
To me, looks good, with some nits inlined.

File: src/java.compiler/share/classes/javax/lang/model/util/Elements.java (hunk 1 above)
Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/util/Elements.java]: Nits fixed; thanks for the review.

**Pull request closed without merging**